File: lib/xevent.ml
open C.Functions

type xid
type time
type atom
type colormap
type drawable
type circulationRequest = PlaceOnTop | PlaceOnBottom
type propertyNotification = PropertyNewValue | PropertyDelete
type colorMapNotification = ColormapUninstalled | ColormapInstalled
type windowStackingMethod = Above | Below | TopIf | BottomIf | Opposite

type message_data =
  | Bytes of char array
  | Shorts of int array
  | Longs of int array

type notifyMode =
  | NotifyNormal
  | NotifyGrab
  | NotifyUngrab
  | NotifyWhileGrabbed

type notifyDetail =
  | NotifyAncestor
  | NotifyVirtual
  | NotifyInferior
  | NotifyNonlinear
  | NotifyNonlinearVirtual
  | NotifyPointer
  | NotifyPointerRoot
  | NotifyDetailNone

type keyEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  root : window;
  subwindow : window;
  time : time;
  pos : int * int;
  root_pos : int * int;
  state : int;
  keycode : int;
  same_screen : bool;
}

type buttonEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  root : window;
  subwindow : window;
  time : time;
  pos : int * int;
  root_pos : int * int;
  state : int;
  button : int;
  same_screen : bool;
}

type buttonPressedEvent = buttonEvent
type buttonReleasedEvent = buttonEvent

type motionEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  root : window;
  subwindow : window;
  time : time;
  pos : int * int;
  root_pos : int * int;
  state : int;
  same_screen : bool;
}

type pointerMovedEvent = motionEvent

type crossingEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  root : window;
  subwindow : window;
  time : time;
  pos : int * int;
  root_pos : int * int;
  mode : notifyMode;
  detail : notifyDetail;
}

type enterWindowEvent = crossingEvent
type leaveWindowEvent = crossingEvent

type focusChangeEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  mode : notifyMode;
  detail : notifyDetail;
}

type focusInEvent = focusChangeEvent
type focusOutEvent = focusChangeEvent

type keymapEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  key_vector : string;
}

type exposeEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  pos : int * int;
  size : int * int;
  count : int;
}

type graphicsExposeEvent = {
  serial : int;
  send_event : bool;
  display : display;
  drawable : drawable;
  pos : int * int;
  count : int;
  codes : int * int;
}

type noExposeEvent = {
  serial : int;
  send_event : bool;
  display : display;
  drawable : drawable;
  codes : int * int;
}

type visibilityEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  state : int;
}

type createWindowEvent = {
  serial : int;
  send_event : bool;
  display : display;
  parent : window;
  window : window;
  pos : int * int;
  size : int * int;
  border_width : int;
  override_redirect : bool;
}

type destroyWindowEvent = {
  serial : int;
  send_event : bool;
  display : display;
  event : window;
  window : window;
}

type unmapEvent = {
  serial : int;
  send_event : bool;
  display : display;
  event : window;
  window : window;
  from_configure : bool;
}

type mapEvent = {
  serial : int;
  send_event : bool;
  display : display;
  event : window;
  window : window;
  override_redirect : bool;
}

type mapRequestEvent = {
  serial : int;
  send_event : bool;
  display : display;
  parent : window;
  window : window;
}

type reparentEvent = {
  serial : int;
  send_event : bool;
  display : display;
  event : window;
  window : window;
  parent : window;
  pos : int * int;
  override_redirect : bool;
}

type configureEvent = {
  serial : int;
  send_event : bool;
  display : display;
  event : window;
  window : window;
  pos : int * int;
  size : int * int;
  border_width : int;
  above : window;
  override_redirect : bool;
}

type gravityEvent = {
  serial : int;
  send_event : bool;
  display : display;
  event : window;
  window : window;
  pos : int * int;
}

type resizeRequestEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  size : int * int;
}

type configureRequestEvent = {
  serial : int;
  send_event : bool;
  display : display;
  parent : window;
  window : window;
  pos : int * int;
  size : int * int;
  border_width : int;
  above : window;
  detail : windowStackingMethod;
  value_mask : int;
}

type circulateEvent = {
  serial : int;
  send_event : bool;
  display : display;
  event : window;
  window : window;
  place : circulationRequest;
}

type circulateRequestEvent = {
  serial : int;
  send_event : bool;
  display : display;
  parent : window;
  window : window;
  place : circulationRequest;
}

type propertyEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  atom : atom;
  time : time;
  state : propertyNotification;
}

type selectionClearEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  selection : atom;
  time : time;
}

type selectionRequestEvent = {
  serial : int;
  send_event : bool;
  display : display;
  owner : window;
  requestor : window;
  selection : atom;
  target : atom;
  property : atom;
  time : time;
}

type selectionEvent = {
  serial : int;
  send_event : bool;
  display : display;
  requestor : window;
  selection : atom;
  target : atom;
  property : atom;
  time : time;
}

type colormapEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  colormap : colormap;
  c_new : bool;
  state : colorMapNotification;
}

type clientMessageEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  message_type : atom;
  data : message_data;
}

type mappingEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
  request : int;
  first_keycode : int;
  count : int;
}

type errorEvent = {
  display : display;
  resourceid : xid;
  serial : int;
  codes : int * int * int;
}

type anyEvent = {
  serial : int;
  send_event : bool;
  display : display;
  window : window;
}

type genericEvent = {
  serial : int;
  send_event : bool;
  display : display;
  extension : int;
  evtype : int;
}

type genericEventCookie = {
  serial : int;
  send_event : bool;
  display : display;
  extension : int;
  evtype : int;
  cookie : int;
  data : string;
}

type xEvent =
  | ErrorEvent of errorEvent (* 0 *)
  | TODO of unit (* 1 *)
  | KeyPress of keyEvent (* 2 *)
  | KeyRelease of keyEvent (* 3 *)
  | ButtonPress of buttonEvent (* 4 *)
  | ButtonRelease of buttonEvent (* 5 *)
  | MotionNotify of motionEvent (* 6 *)
  | EnterNotify of crossingEvent (* 7 *)
  | LeaveNotify of crossingEvent (* 8 *)
  | FocusIn of focusChangeEvent (* 9 *)
  | FocusOut of focusChangeEvent (* 10 *)
  | KeymapNotify of keymapEvent (* 11 *)
  | Expose of exposeEvent (* 12 *)
  | GraphicsExpose of graphicsExposeEvent (* 13 *)
  | NoExpose of noExposeEvent (* 14 *)
  | VisibilityNotify of visibilityEvent (* 15 *)
  | CreateWindow of createWindowEvent (* 16 *)
  | DestroyWindow of destroyWindowEvent (* 17 *)
  | Unmap of unmapEvent (* 18 *)
  | Map of mapEvent (* 19 *)
  | MapRequest of mapRequestEvent (* 20 *)
  | Reparent of reparentEvent (* 21 *)
  | ConfigureNotify of configureEvent (* 22 *)
  | ConfigureRequest of configureRequestEvent (* 23 *)
  | GravityNotify of gravityEvent (* 24 *)
  | ResizeRequest of resizeRequestEvent (* 25 *)
  | Circulate of circulateEvent (* 26 *)
  | CirculateRequest of circulateRequestEvent (* 27 *)
  | Property of propertyEvent (* 28 *)
  | SelectionClear of selectionClearEvent (* 29 *)
  | SelectionRequest of selectionRequestEvent (* 30 *)
  | Selection of selectionEvent (* 31 *)
  | Colormap of colormapEvent (* 32 *)
  | ClientMessage of clientMessageEvent (* 33 *)
  | Mapping of mappingEvent (* 34 *)
  | Generic of genericEvent (* 35 *)
  | GenericCookie of genericEventCookie (* 36 *)

File: lib/xevent.c
#include "utils.h"
#include <X11/Xlib.h>
#include <caml/alloc.h>
#include <caml/fail.h>
#include <caml/memory.h>
#include <caml/mlvalues.h>

value Val_xevent(XEvent *event) {
  CAMLparam0();
  CAMLlocal5(v_record, v_pos, v_root_pos, v_size, v_event);
  int off = 0;

  switch (event->type) {
  case KeyPress:
  case KeyRelease:
    v_record = caml_alloc(12, 0); // 12 fields
    Store_field(v_record, off++, Val_int(event->xkey.serial));
    Store_field(v_record, off++, Val_bool(event->xkey.send_event));
    Store_field(v_record, off++, Val_ptr(event->xkey.display));
    Store_field(v_record, off++, Val_xid(event->xkey.window));
    Store_field(v_record, off++, Val_xid(event->xkey.root));
    Store_field(v_record, off++, Val_xid(event->xkey.subwindow));
    Store_field(v_record, off++, Val_xid(event->xkey.time));
    v_pos = caml_tuple2(Val_int(event->xkey.x), Val_int(event->xkey.y));
    Store_field(v_record, off++, v_pos);
    v_root_pos =
        caml_tuple2(Val_int(event->xkey.x_root), Val_int(event->xkey.y_root));
    Store_field(v_record, off++, v_root_pos);
    Store_field(v_record, off++, Val_int(event->xkey.state));
    Store_field(v_record, off++, Val_int(event->xkey.keycode));
    Store_field(v_record, off++, Val_bool(event->xkey.same_screen));
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  case ButtonPress:
  case ButtonRelease:
    v_record = caml_alloc(12, 0); //  12 fields
    Store_field(v_record, off++, Val_int(event->xbutton.serial));
    Store_field(v_record, off++, Val_bool(event->xbutton.send_event));
    Store_field(v_record, off++, Val_ptr(event->xbutton.display));
    Store_field(v_record, off++, Val_xid(event->xbutton.window));
    Store_field(v_record, off++, Val_xid(event->xbutton.root));
    Store_field(v_record, off++, Val_xid(event->xbutton.subwindow));
    Store_field(v_record, off++, Val_xid(event->xbutton.time));
    v_pos = caml_tuple2(Val_int(event->xbutton.x), Val_int(event->xbutton.y));
    Store_field(v_record, off++, v_pos);
    v_root_pos = caml_tuple2(Val_int(event->xbutton.x_root),
                             Val_int(event->xbutton.y_root));
    Store_field(v_record, off++, v_root_pos);
    Store_field(v_record, off++, Val_int(event->xbutton.state));
    Store_field(v_record, off++, Val_int(event->xbutton.button));
    Store_field(v_record, off++, Val_bool(event->xbutton.same_screen));
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  case MotionNotify:
    v_record = caml_alloc(13, 0); // 13 fields
    Store_field(v_record, off++, Val_int(event->xmotion.serial));
    Store_field(v_record, off++, Val_bool(event->xmotion.send_event));
    Store_field(v_record, off++, Val_ptr(event->xmotion.display));
    Store_field(v_record, off++, Val_xid(event->xmotion.window));
    Store_field(v_record, off++, Val_xid(event->xmotion.root));
    Store_field(v_record, off++, Val_xid(event->xmotion.subwindow));
    Store_field(v_record, off++, Val_xid(event->xmotion.time));
    v_pos = caml_tuple2(Val_int(event->xmotion.x), Val_int(event->xmotion.y));
    Store_field(v_record, off++, v_pos);
    v_root_pos = caml_tuple2(Val_int(event->xmotion.x_root),
                             Val_int(event->xmotion.y_root));
    Store_field(v_record, off++, v_root_pos);
    Store_field(v_record, off++, Val_int(event->xmotion.state));
    Store_field(v_record, off++, Val_bool(event->xmotion.same_screen));
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  case Expose:
    v_record = caml_alloc(7, 0); // 7 fields
    Store_field(v_record, off++, Val_int(event->xexpose.serial));
    Store_field(v_record, off++, Val_bool(event->xexpose.send_event));
    Store_field(v_record, off++, Val_ptr(event->xexpose.display));
    Store_field(v_record, off++, Val_xid(event->xexpose.window));
    v_pos = caml_tuple2(Val_int(event->xexpose.x), Val_int(event->xexpose.y));
    Store_field(v_record, off++, v_pos);
    v_size = caml_tuple2(Val_int(event->xexpose.width),
                         Val_int(event->xexpose.height));
    Store_field(v_record, off++, v_size);
    Store_field(v_record, off++, Val_int(event->xexpose.count));
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  case DestroyNotify:
    v_record = caml_alloc(5, 0);
    Store_field(v_record, off++, Val_int(event->xdestroywindow.serial));
    Store_field(v_record, off++, Val_bool(event->xdestroywindow.send_event));
    Store_field(v_record, off++, Val_ptr(event->xdestroywindow.display));
    Store_field(v_record, off++, Val_xid(event->xdestroywindow.event));
    Store_field(v_record, off++, Val_xid(event->xdestroywindow.window));
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  case UnmapNotify:
    v_record = caml_alloc(6, 0);
    Store_field(v_record, off++, Val_int(event->xunmap.serial));
    Store_field(v_record, off++, Val_bool(event->xunmap.send_event));
    Store_field(v_record, off++, Val_ptr(event->xunmap.display));
    Store_field(v_record, off++, Val_xid(event->xunmap.event));
    Store_field(v_record, off++, Val_xid(event->xunmap.window));
    Store_field(v_record, off++, Val_bool(event->xunmap.from_configure));
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  case MapNotify:
    v_record = caml_alloc(6, 0);
    Store_field(v_record, off++, Val_int(event->xmap.serial));
    Store_field(v_record, off++, Val_bool(event->xmap.send_event));
    Store_field(v_record, off++, Val_ptr(event->xmap.display));
    Store_field(v_record, off++, Val_xid(event->xmap.event));
    Store_field(v_record, off++, Val_xid(event->xmap.window));
    Store_field(v_record, off++, Val_bool(event->xmap.override_redirect));
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  case MapRequest:
    v_record = caml_alloc(5, 0);
    Store_field(v_record, off++, Val_int(event->xmaprequest.serial));
    Store_field(v_record, off++, Val_bool(event->xmaprequest.send_event));
    Store_field(v_record, off++, Val_ptr(event->xmaprequest.display));
    Store_field(v_record, off++, Val_xid(event->xmaprequest.parent));
    Store_field(v_record, off++, Val_xid(event->xmaprequest.window));
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  case ConfigureRequest:
    v_record = caml_alloc(12, 0);
    Store_field(v_record, off++, Val_int(event->xconfigurerequest.serial));
    Store_field(v_record, off++, Val_bool(event->xconfigurerequest.send_event));
    Store_field(v_record, off++, Val_ptr(event->xconfigurerequest.display));
    Store_field(v_record, off++, Val_xid(event->xconfigurerequest.parent));
    Store_field(v_record, off++, Val_xid(event->xconfigurerequest.window));
    v_pos = caml_tuple2(Val_int(event->xconfigurerequest.x),
                        Val_int(event->xconfigurerequest.y));
    Store_field(v_record, off++, v_pos);
    v_size = caml_tuple2(Val_int(event->xconfigurerequest.width),
                         Val_int(event->xconfigurerequest.height));
    Store_field(v_record, off++, v_size);
    Store_field(v_record, off++,
                Val_int(event->xconfigurerequest.border_width));
    Store_field(v_record, off++, Val_xid(event->xconfigurerequest.above));
    Store_field(v_record, off++, Val_int(event->xconfigurerequest.detail));
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  case ReparentNotify:
    v_record = caml_alloc(9, 0); // 9 fields
    Store_field(v_record, off++, Val_int(event->xreparent.serial));
    Store_field(v_record, off++, Val_bool(event->xreparent.send_event));
    Store_field(v_record, off++, Val_ptr(event->xreparent.display));
    Store_field(v_record, off++, Val_xid(event->xreparent.event));
    Store_field(v_record, off++, Val_xid(event->xreparent.window));
    Store_field(v_record, off++, Val_xid(event->xreparent.parent));
    v_pos =
        caml_tuple2(Val_int(event->xreparent.x), Val_int(event->xreparent.y));
    Store_field(v_record, off++, v_pos);
    Store_field(v_record, off++, Val_bool(event->xreparent.override_redirect));
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  case ConfigureNotify:
    v_record = caml_alloc(12, 0); // 12 fields
    Store_field(v_record, off++, Val_int(event->xconfigure.serial));
    Store_field(v_record, off++, Val_bool(event->xconfigure.send_event));
    Store_field(v_record, off++, Val_ptr(event->xconfigure.display));
    Store_field(v_record, off++, Val_xid(event->xconfigure.event));
    Store_field(v_record, off++, Val_xid(event->xconfigure.window));
    v_pos =
        caml_tuple2(Val_int(event->xconfigure.x), Val_int(event->xconfigure.y));
    Store_field(v_record, off++, v_pos);
    v_size = caml_tuple2(Val_int(event->xconfigure.width),
                         Val_int(event->xconfigure.height));
    Store_field(v_record, off++, v_size);
    Store_field(v_record, off++, Val_int(event->xconfigure.border_width));
    Store_field(v_record, off++, Val_xid(event->xconfigure.above));
    Store_field(v_record, off++, Val_bool(event->xconfigure.override_redirect));
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  case PropertyNotify:
    v_record = caml_alloc(7, 0); // 7 fields
    Store_field(v_record, off++, Val_int(event->xproperty.serial));
    Store_field(v_record, off++, Val_bool(event->xproperty.send_event));
    Store_field(v_record, off++, Val_ptr(event->xproperty.display));
    Store_field(v_record, off++, Val_xid(event->xproperty.window));
    Store_field(v_record, off++, Val_xid(event->xproperty.atom));
    Store_field(v_record, off++, Val_xid(event->xproperty.time));
    Store_field(v_record, off++, Val_int(event->xproperty.state));
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  case ClientMessage:
    CAMLlocal2(v_data, v_array);
    switch (event->xclient.format) {
    case 8:
      v_array = caml_alloc(20, 0); // tag 0 for standard OCaml array
      for (int i = 0; i < 20; i++) {
        Store_field(v_array, i, Val_int(event->xclient.data.b[i]));
      }
      v_data = caml_alloc(1, 0); // tag 0 for Bytes
      Store_field(v_data, 0, v_array);
      break;
    case 16:
      v_array = caml_alloc(10, 0); // tag 0 for standard OCaml array
      for (int i = 0; i < 10; i++) {
        Store_field(v_array, i, Val_int(event->xclient.data.s[i]));
      }
      v_data = caml_alloc(1, 1); // tag 1 for Shorts
      Store_field(v_data, 0, v_array);
      break;
    case 32:
      v_array = caml_alloc(5, 0); // tag 0 for standard OCaml array
      for (int i = 0; i < 5; i++) {
        Store_field(v_array, i, Val_int(event->xclient.data.l[i]));
      }
      v_data = caml_alloc(1, 2); // tag 2 for Longs
      Store_field(v_data, 0, v_array);
      break;

    default:
      caml_failwith_fmt("Invalid ClientMessage format: %d",
                        event->xclient.format);
    }

    v_record = caml_alloc(6, 0); // 7 fields
    Store_field(v_record, off++, Val_int(event->xclient.serial));
    Store_field(v_record, off++, Val_bool(event->xclient.send_event));
    Store_field(v_record, off++, Val_ptr(event->xclient.display));
    Store_field(v_record, off++, Val_xid(event->xclient.window));
    Store_field(v_record, off++, Val_xid(event->xclient.message_type));
    Store_field(v_record, off++, v_data);
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  case FocusIn:
    v_record = caml_alloc(6, 0); // 6 fields
    Store_field(v_record, off++, Val_int(event->xfocus.serial));
    Store_field(v_record, off++, Val_bool(event->xfocus.send_event));
    Store_field(v_record, off++, Val_ptr(event->xfocus.display));
    Store_field(v_record, off++, Val_xid(event->xfocus.window));
    Store_field(v_record, off++, Val_int(event->xfocus.mode));
    Store_field(v_record, off++, Val_int(event->xfocus.detail));
    v_event = caml_alloc(1, event->type);
    Store_field(v_event, 0, v_record);
    break;

  default:
    caml_failwith_fmt("Unsupported event type %d", event->type);
  }
  CAMLreturn(v_event);
}
